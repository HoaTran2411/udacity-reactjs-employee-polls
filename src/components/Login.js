import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { setAuthedUser } from '../actions/users';
import { handleInitialData } from '../actions/share';
import { _getUsers } from '../utils/_DATA';

const Login = () => {
  const navigate = useNavigate();
  const dispatch = useDispatch();
  const [users, setUsers] = useState([]);
  const [userId, setUserId] = useState('');

  useEffect(() => {
    _getUsers().then(setUsers);
  }, []);

  const handleLogin = () => {
    if (userId) {
      dispatch(setAuthedUser(userId));
      dispatch(handleInitialData());
      navigate('/');
    }
  };

  return (
    <div>
      <h1>Login</h1>
      <select className='inputText' onChange={(e) => setUserId(e.target.value)} value={userId}>
        <option value="">Select User</option>
        {Object.keys(users).map((id) => (
          <option key={id} value={id}>{users[id].name}</option>
        ))}
      </select>
      <button onClick={handleLogin} className='btn'>Login now</button>
    </div>
  );
};

export default Login;