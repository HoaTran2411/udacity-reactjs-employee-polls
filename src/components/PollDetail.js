import React from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { useParams } from 'react-router-dom';
import { handleAnswerPoll } from '../actions/share';
import '../App.css';

const PollDetail = () => {
  const dispatch = useDispatch();
  const { id: pollId } = useParams();
  const poll = useSelector((state) => state.polls[pollId]);
  const authedUser = useSelector((state) => state.users.authedUser);
  const users = useSelector((state) => state.users);

  if (authedUser == null) {
    return (<div>Please login first</div>);
  }

  if (!poll) {
    return <div>404 Poll Not Found</div>;
  }

  // Handle number and percentage of people who voted option
  const numberVotedOption1 = poll.optionOne.votes.length;
  const numberVotedOption2 = poll.optionTwo.votes.length;
  const numberOfUsers = Object.keys(users).length - 1;
  const percentageVotedOption1 = numberVotedOption1 / numberOfUsers * 100;
  const percentageVotedOption2 = numberVotedOption2 / numberOfUsers * 100;

  let answerPoll = null;
  if (poll?.optionOne.votes.includes(authedUser)) {
    answerPoll = 'optionOne';
  } else if (poll?.optionTwo.votes.includes(authedUser)) {
    answerPoll = 'optionTwo';
  }

  const handleVote = (answer) => {
    answerPoll = answer;
    dispatch(handleAnswerPoll(pollId, answer, authedUser));
  };

  return (
    <div>
      <p>Author: {poll.author}</p>
      <img className='userImg' src={users[poll.author].avatarURL} alt='avatar' />
      <h2>Would You Rather</h2>
      <h3>{poll.optionOne.text}
        <span> ({numberVotedOption1} voted ({percentageVotedOption1}%)) </span>
        <span className="compareText">or </span>
        {poll.optionTwo.text}
        <span> ({numberVotedOption2} voted ({percentageVotedOption2}%)) </span>
      </h3>
      {!answerPoll &&
        <div>
          <button onClick={() => handleVote('optionOne')} className='btn'>
            Vote for Option One
          </button>
          <button onClick={() => handleVote('optionTwo')} className='btn'>
            Vote for Option Two </button>
        </div>
      }
      {answerPoll && <p>You voted for {answerPoll}</p>}
      {!answerPoll && <p>You not yet voted, please vote!</p>}
    </div >
  );
};

export default PollDetail;