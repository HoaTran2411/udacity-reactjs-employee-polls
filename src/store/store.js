import { createStore, combineReducers } from 'redux';
import polls from '../reducers/polls';
import middleware from '../middleware'
import users from '../reducers/users';

export const rootReducer = combineReducers({
  polls,
  users,
});

const store = createStore(rootReducer, middleware);

export default store;