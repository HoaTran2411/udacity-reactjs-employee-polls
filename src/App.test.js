import React from 'react';
import '@testing-library/jest-dom';
import { Provider } from 'react-redux';
import { rootReducer } from './store/store';
import { createStore, applyMiddleware } from 'redux';
import { thunk } from 'redux-thunk';
import { render, fireEvent, screen } from '@testing-library/react';
import PollDetail from './components/PollDetail';
import Login from './components/Login';
import { MemoryRouter, Route, Routes } from 'react-router-dom';
import { _saveQuestion, _saveQuestionAnswer } from './utils/_DATA.js';

const mockStore = (initialState) => {
  return createStore(rootReducer, initialState, applyMiddleware(thunk));
};

let users;
let questions;

describe('Check component <Login/>', () => {
  let store;
  beforeEach(() => {
    store = mockStore({});
    jest.clearAllMocks();
  });

  test('dispatches action and navigates on button click', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/']}>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    const button = screen.getByText('Login now');
    fireEvent.click(button);
    expect(window.location.pathname).toBe('/');
  });

  test('Check text in screen', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={['/']}>
          <Login />
        </MemoryRouter>
      </Provider>
    );

    expect(screen.getByText('Select User')).toBeInTheDocument();
  });
});

describe('<PollDetail />', () => {
  let store;
  const id = '8xf0y6ziyjabvozdd253nd';
  const poll = {
    id: id,
    author: 'sarahedo',
    timestamp: 1467166872634,
    optionOne: {
      votes: [],
      text: 'Build our new application with Javascript',
    },
    optionTwo: {
      votes: [],
      text: 'Build our new application with Typescript'
    }
  };

  beforeEach(() => {
    store = mockStore({
      users: {
        authedUser: 'sarahedo',
        sarahedo: {
          id: 'sarahedo',
          password: 'password123',
          name: 'Sarah Edo',
          avatarURL: 'https://img.freepik.com/premium-photo/engaging-cartoon-avatar-trustworthy-girl-her-late-s-tiktok-promotions_1283595-3611.jpg',
          answers: {
            "8xf0y6ziyjabvozdd253nd": 'optionOne',
            "6ni6ok3ym7mf1p33lnez": 'optionOne',
            "am8ehyc8byjqgar0jgpub9": 'optionTwo',
            "loxhs1bqm25b708cmbf3g": 'optionTwo'
          },
          questions: ['8xf0y6ziyjabvozdd253nd', 'am8ehyc8byjqgar0jgpub9']
        },
      },
      polls: {
        [id]: poll,
      },
    });
  });

  test('renders poll', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/poll/${id}`]}>
          <Routes>
            <Route path="/poll/:id" element={<PollDetail />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByText(/Would You Rather/i)).toBeInTheDocument();
  });

  test('check exist option 1', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/poll/${id}`]}>
          <Routes>
            <Route path="/poll/:id" element={<PollDetail />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByText(/Vote for Option One/i)).toBeInTheDocument();
  });

  test('check exist option 2', () => {
    render(
      <Provider store={store}>
        <MemoryRouter initialEntries={[`/poll/${id}`]}>
          <Routes>
            <Route path="/poll/:id" element={<PollDetail />} />
          </Routes>
        </MemoryRouter>
      </Provider>
    );
    expect(screen.getByText(/Vote for Option One/i)).toBeInTheDocument();
  });
});

// Test the _saveQuestion() function in _DATA.js
test('_saveQuestion should save data correctly', async () => {
  const poll = {
    author: 'sarahedo',
    optionOneText: 'Option A',
    optionTwoText: 'Option B',
  };

  const savedPoll = await _saveQuestion(poll);
  expect(savedPoll.optionOne.text).toBe(poll.optionOneText);
  expect(savedPoll.optionTwo.text).toBe(poll.optionTwoText);
  expect(savedPoll.author).toBe(poll.author);
});

test('check _saveQuestion when return an error', async () => {
  const question = { author: null };
  await expect(_saveQuestion(question)).rejects.toEqual('OptionOneText, optionTwoText, and author must not null');
});

// Test the _saveQuestionAnswer() function in _DATA.js
test('check _saveQuestionAnswer return correct when data is correct', async () => {
  const result = await _saveQuestionAnswer({ authedUser: 'tylermcginnis', qid: '6ni6ok3ym7mf1p33lnez', answer: 'optionOne' });
  expect(result).toBe(undefined);
});

// An async unit test to verify that an error is returned if incorrect data is passed to the function
test('check _saveQuestionAnswer when input data is error', async () => {
  const authedUser = null;
  const qid = 'abcd1234';
  const answer = 'invalidOption';
  await expect(_saveQuestionAnswer({ authedUser, qid, answer })).rejects.toEqual('authedUser must not null');
});

describe('check function _saveQuestionAnswer', () => {
  beforeEach(() => {
    users = {
      sarahedo: {
        id: 'sarahedo',
        password: 'password123',
        name: 'Sarah Edo',
        avatarURL: 'https://img.freepik.com/premium-photo/engaging-cartoon-avatar-trustworthy-girl-her-late-s-tiktok-promotions_1283595-3611.jpg',
        answers: {
          "8xf0y6ziyjabvozdd253nd": 'optionOne',
        },
        questions: ['8xf0y6ziyjabvozdd253nd']
      },
      tylermcginnis: {
        id: 'tylermcginnis',
        password: 'abc321',
        name: 'Tyler McGinnis',
        avatarURL: 'https://gravatar.com/avatar/27205e5c51cb03f862138b22bcb5dc20f94a342e744ff6df1b8dc8af3c865109',
        answers: {
        },
        questions: [],
      }
    };

    questions = {
      "8xf0y6ziyjabvozdd253nd": {
        id: '8xf0y6ziyjabvozdd253nd',
        author: 'sarahedo',
        timestamp: 1467166872634,
        optionOne: {
          votes: ['sarahedo'],
          text: 'Build our new application with Javascript',
        },
        optionTwo: {
          votes: [],
          text: 'Build our new application with Typescript'
        }
      },
    };
  });

  test('check snapshot is correct for _saveQuestionAnswer', async () => {
    await _saveQuestionAnswer({
      authedUser: 'tylermcginnis',
      qid: '8xf0y6ziyjabvozdd253nd',
      answer: 'optionOne'
    });

    expect(users).toMatchSnapshot();
    expect(questions).toMatchSnapshot();
  });
});