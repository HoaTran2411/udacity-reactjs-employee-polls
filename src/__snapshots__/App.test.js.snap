// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`_saveQuestionAnswer should update the users and questions state correctly 1`] = `
Object {
  "sarahedo": Object {
    "answers": Object {
      "8xf0y6ziyjabvozdd253nd": "optionOne",
    },
    "avatarURL": "https://img.freepik.com/premium-photo/engaging-cartoon-avatar-trustworthy-girl-her-late-s-tiktok-promotions_1283595-3611.jpg",
    "id": "sarahedo",
    "name": "Sarah Edo",
    "password": "password123",
    "questions": Array [
      "8xf0y6ziyjabvozdd253nd",
    ],
  },
  "tylermcginnis": Object {
    "answers": Object {},
    "avatarURL": "https://gravatar.com/avatar/27205e5c51cb03f862138b22bcb5dc20f94a342e744ff6df1b8dc8af3c865109",
    "id": "tylermcginnis",
    "name": "Tyler McGinnis",
    "password": "abc321",
    "questions": Array [],
  },
}
`;

exports[`_saveQuestionAnswer should update the users and questions state correctly 2`] = `
Object {
  "8xf0y6ziyjabvozdd253nd": Object {
    "author": "sarahedo",
    "id": "8xf0y6ziyjabvozdd253nd",
    "optionOne": Object {
      "text": "Build our new application with Javascript",
      "votes": Array [
        "sarahedo",
      ],
    },
    "optionTwo": Object {
      "text": "Build our new application with Typescript",
      "votes": Array [],
    },
    "timestamp": 1467166872634,
  },
}
`;

exports[`check function _saveQuestionAnswer check snapshot is correct for _saveQuestionAnswer 1`] = `
Object {
  "sarahedo": Object {
    "answers": Object {
      "8xf0y6ziyjabvozdd253nd": "optionOne",
    },
    "avatarURL": "https://img.freepik.com/premium-photo/engaging-cartoon-avatar-trustworthy-girl-her-late-s-tiktok-promotions_1283595-3611.jpg",
    "id": "sarahedo",
    "name": "Sarah Edo",
    "password": "password123",
    "questions": Array [
      "8xf0y6ziyjabvozdd253nd",
    ],
  },
  "tylermcginnis": Object {
    "answers": Object {},
    "avatarURL": "https://gravatar.com/avatar/27205e5c51cb03f862138b22bcb5dc20f94a342e744ff6df1b8dc8af3c865109",
    "id": "tylermcginnis",
    "name": "Tyler McGinnis",
    "password": "abc321",
    "questions": Array [],
  },
}
`;

exports[`check function _saveQuestionAnswer check snapshot is correct for _saveQuestionAnswer 2`] = `
Object {
  "8xf0y6ziyjabvozdd253nd": Object {
    "author": "sarahedo",
    "id": "8xf0y6ziyjabvozdd253nd",
    "optionOne": Object {
      "text": "Build our new application with Javascript",
      "votes": Array [
        "sarahedo",
      ],
    },
    "optionTwo": Object {
      "text": "Build our new application with Typescript",
      "votes": Array [],
    },
    "timestamp": 1467166872634,
  },
}
`;

exports[`should update the users and questions state correctly 1`] = `undefined`;

exports[`should update the users and questions state correctly 2`] = `undefined`;
